### ISSUE TITLE
[graphql-lsp] Template string fragment get `Unknown fragment` error

### ISSUE THREAD

**Issue by VinceBT-BG** [graphql/graphiql#3066]
When inserting a fragment inside a template string, the fragment does not get recognized when it gets called in the rest of the gql string

![2023-03-01 11 27 00](https://user-images.githubusercontent.com/117643868/222113242-a0cfdcfb-8c76-47e6-9686-e8aa6b0971c0.gif)

IDE: VSCode `1.75.1`

Extensions: 
- GraphQL: Language Feature Support `v0.8.5`
- GraphQL: Syntax Highlighting `v1.0.6`

**Comment by avlaguta** [comment]
The related issue:
https://github.com/graphql/graphiql/issues/2064

**Comment by acao** [comment]
there must have been a change that broke this use case. it works if the fragments are in a separate file, or are defined in the same graphql string, which is what the tests cover

**Comment by acao** [comment]
Huh, I can't seem to reproduce the bug now, is it fixed for you in the latest version?

![fragment-validation-bug](https://github.com/graphql/graphiql/assets/1368727/8e2791fc-660a-49cf-9ac7-b789de7de27c)

**Comment by VinceBT-BG** [comment]
Can you show your config and schema ? Here is my config, I am using a different root if that helps
![2023-06-26 14 22 32](https://github.com/graphql/graphiql/assets/117643868/a459898c-6d2e-49e5-a387-a073ccfad863)

**Comment by VinceBT-BG** [comment]
I tried to reproduce your example and I don't have the same result, can you show your config/schema ?

![2023-06-26 15 49 31](https://github.com/graphql/graphiql/assets/117643868/0cb0f438-4f82-4369-be15-6f8468faed1e)

**Comment by acao** [comment]
@VinceBT-BG here is the config;

```ts
export default {
   schema: "src/schema/schema.graphql",
   documents: "src/queries/**/*.ts",
};
```
```graphql
# for example
type ExampleType {
   shelter: String
   something: String
   however: String
}

type Query {
    exampleField: ExampleType
    love: String
    message: String
}
```

**Comment by VinceBT-BG** [comment]
I was indeed using `schemaPath` instead of `schema` in the configuration! I tried on the example and it worked, I also tried in my project and I don't have the errors anymore, thanks !

**Comment by acao** [comment]
@VinceBT-BG music to my ears! another happy user. I will add something to the FAQ section for users using legacy `schemaPath`

awesome gif tool by the way, i like the keyboard icons, maybe I will use this!

**Comment by Crunchyman-ralph** [comment]
I'm still having this issue:
![2023-09-28 22 10 07](https://github.com/graphql/graphiql/assets/35776126/507968cc-784e-4afe-a194-58738ec08575)

fragment file:
```ts
import { gql } from '@sb/webapp-api-client/graphql';

export const processTestItemListQuery = gql(/* GraphQL */ `
  query processTestItemListQuery {
    allProcessTestItems(first: 100) {
      edges {
        node {
          ...processTestItemListItemFragment
        }
      }
    }
  }
`);

export const processTestItemListItemDeleteMutation = gql(/* GraphQL */ `
  mutation processTestItemListItemDeleteMutation($input: DeleteProcessTestItemMutationInput!) {
    deleteProcessTestItem(input: $input) {
      deletedIds
    }
  }
`);

export const processTestItemListItemFragment = gql(/* GraphQL */ `
  fragment processTestItemListItemFragment on ProcessTestItemType {
    id
    name
    price
  }
`);
```

.graphqlconfig:
```json
{
  "name": "Untitled GraphQL Schema",
  "projects": {
    "api": {
      "name": "GraphQL API",
      "schema": "packages/webapp-libs/webapp-api-client/graphql/schema/api.graphql"
    },
    "contentful": {
      "name": "Contentful API",
      "schema": "packages/webapp-libs/webapp-contentful/graphql/schema/contentful.graphql"
    }
  },
  "documents": "**/*.ts"
}
```

**Comment by ypresto** [comment]
@acao I still have same issue now. Should I create new issue or could you reopen this issue? :pray:

**Comment by acao** [comment]
@ypresto can you open a new issue, and please provide the config? if it's a multi project config then it's a different bug with the same message

**Comment by sachingaikwad123** [comment]
I am still facing this issue in my vscode. Here is my graphql-config:

{
  "schema": "http://127.0.0.1:8000/graphql/",
  "documents": ["src/**/*.{ts,tsx}"]
}

My fragment is defined in another file.

Q: Is there any solution for this?

**Comment by acao** [comment]
@sachingaikwad123 can you see if the pre-release in #3521 fixes the issue? we have tests that I think should confirm this issue is resolved.

what I discovered was we were updating the fragment cache for onUpdate and onWatchedFilesUpdate, but not on onOpenOrAdd, and some other issues with fragment cache keys that piled up along the way. now we have an integration test suite for the server that confirms a happy path for validating fragments and fragments on queries across changes to all files. i will double check to make sure we are still confirming validation is empty where it should be on every mutative event

**Comment by acao** [comment]
The only thing i want to clarify here with the incoming fix is that we are not inferring information from template strings. If a fragment is present in any file in the project, it can be used in any query, and it's up to the client framework to decide how to resolve fragments

**Comment by xkelxmc** [comment]
Same problem with using separate files

**Comment by efstajas** [comment]
<img width="1817" alt="image" src="https://github.com/graphql/graphiql/assets/1018218/ce563e5f-e93a-404f-bba4-eec156af48f5">

I'm running into the same issue now. I have a lot of fragments defined within `.svelte` files in a "context="module"" script block, but I don't think this should make a difference. I include all `.ts` and `.svelte` files in my `documents`. Here's my config:

```
schema: 'src/lib/graphql/__generated__/schema.graphql'
documents: 'src/**/*.{graphql,js,ts,svelte}'
ignoreNoDocuments: true
```

**Comment by efstajas** [comment]
Regarding my previous comment: I've tracked it down to the fragment imports. SvelteKit / TS allow imports to be made either within the main `<script>` tag OR the other script tag with `context="module"`, and access them in either. It seems that when the import is made in a different script tag than that which includes the `gql` definition, it won't see that fragment and treat it as unknown.

**Comment by acao** [comment]
i have a theory this bug has resurfaced because of another schema cacheing bug I've been working on - thus the fragments are "Unknown" because their base type is missing

**Comment by efstajas** [comment]
@acao if it helps — i've noticed since that pretty consistently the error disappears when i:

- save the file that has the schema in question (even without any changes)
- save the file in which the error occurs (also without changes)

after that, it seems like pretty much 100% of the time, the error disappears for all the fragments defined in the 1st file.

**Comment by guneyozsan** [comment]
We have all the fragments in different files. Only one of them has the issue where all have similar structures.

I couldn't notice a pattern how it happens. There were no issue until some point (see downgrade not below). Then some files previously working fine started displaying red marks without any modifications.

- Downgrading from `0.11.0` to `0.9.3` resolves the issue.
- Reinstalling `0.11.0` revives the issue on the same fragment.
- Saving with or without changes does not help.

**Comment by yamatsum** [comment]
@acao Is there a plan to fix this bug?

**Comment by brentshulman-silkline** [comment]
> @acao if it helps — i've noticed since that pretty consistently the error disappears when i:
> 
> * save the file that has the schema in question (even without any changes)
> * save the file in which the error occurs (also without changes)
> 
> after that, it seems like pretty much 100% of the time, the error disappears for all the fragments defined in the 1st file.

Exactly the same for myself

**Comment by acao** [comment]
@yamatsum yes it was one of the goala of the last refactor, but the test cases I resolved for don't seem to cover some of these cases. if someone could provide a test case for MessageProcessor.spec.ts that re-creates this reliably it would be awesome!

**Comment by richard-stafflink** [comment]
A college (MAC) has the same issue, but I don't (Linux) for the same codebase.
Interestingly, some fragments are recognised, some not, so not a consistent problem.
What's more interesting is that one particular fragment is recognised in one file, but not in another - Strange!

**Comment by ya2s** [comment]
@motemen @acao Has this issue already been resolved?

**Comment by motemen** [comment]
@ya2s I don't know. Can't you check it?

**Comment by VinceBT-BG** [comment]
I'll close it, it should have been fixed in https://github.com/graphql/graphiql/pull/3861
